19.0 app traning raibr

ramezlahzy opened the pull request.

Commit:
[ADD] estate: Initial module setup

Commit:
[ADD] estate: Complete Chapter 1 requirements

Commit:
[ADD] estate: Complete Chapter 2 requirements

Commit:
[ADD] estate: Complete Chapter 3 requirements

ramezlahzy committed:
[ADD] estate: Minimum security configuration (Chapter 4)

ramezlahzy committed:
[ADD] estate: Property views and menus (Chapter 5)

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1074.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1074)

ramezlahzy committed:
[ADD] estate: Search filters and grouping options (Chapter 6)

lost-odoo commented:
Hello, can you please ensure that your runbot is green :smile:

ramezlahzy committed:
[FIX] estate: Chapter 7 fixes

ramezlahzy committed:
[FIX] estate: Chapter 7 corrections

ramezlahzy committed:
[ADD] estate: Implement property offers with computed fields and onchange

ramezlahzy committed:
[ADD] estate(Chapter 9): Add property state workflow with sold/cancel actions and offer acceptance

ramezlahzy committed:
[IMP] estate: Add .DS_Store to .gitignore

ramezlahzy committed:
[IMP] estate: Fix ruff linting issues

ramezlahzy commented:
> Hello, can you please ensure that your runbot is green 😄

Can you check again ;)

ramezlahzy committed:
[FIX] estate: replace invalid models.Constraint with _sql_constraints

The models.Constraint syntax does not exist in Odoo's ORM.
SQL constraints must be defined using the _sql_constraints class attribute.

ramezlahzy committed:
[FIX] estate: replace invalid models.Constraint with _sql_constraints

The models.Constraint syntax does not exist in Odoo's ORM.
SQL constraints must be defined using the _sql_constraints class attribute.

lost-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_estate_property,access_estate_property,model_estate_property,base.group_user,1,1,1,1
+access_estate_property_type,access_estate_property_type,model_estate_property_type,base.group_user,1,1,1,1
+access_estate_property_tag,access_estate_property_tag,model_estate_property_tag,base.group_user,1,1,1,1
+access_estate_property_offer,access_estate_property_offer,model_estate_property_offer,base.group_user,1,1,1,1
```
Don't forget to add a final new lines.

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+    status = fields.Selection(
+        selection=[("accepted", "Accepted"), ("refused", "Refused")], copy=False
+    )
+    partner_id = fields.Many2one("res.partner", string="Partner", required=True)
```
It is good practice to name your parameters. So here you should add `comodel_name='res.partner'`.

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class EstatePropertyOffer(models.Model):
+    _name = "estate.property.offer"
```
You should only use double quotes for non technical fields and single quotes for technical fields.

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+
+    price = fields.Float(required=True)
+    _check_price = models.Constraint(
+        "Check(price > 0)", "The offer price must be strictly positive."
```
Splitting this on multiple lines doesn't provide more readability. You can keep this on a single line.

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+            record.date_deadline = fields.Date.today() + timedelta(days=record.validity)
+
+    def _inverse_date_deadline(self):
+        for record in self:
```
Instead of `record` try to name the variable accordingly to the name of the model so something like `property_offer` or `offer` would be better.

Same thing at other places

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+        for record in self:
+            if record.property_id.offer_ids.filtered(
+                lambda o: o.status == "accepted" and o.id != record.id
+            ):
```
It doesn't provide more readability to split this into multiple lines.

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+            if record.property_id.offer_ids.filtered(
+                lambda o: o.status == "accepted" and o.id != record.id
+            ):
+                raise UserError("Only one offer can be accepted per property.")
```
You should make this string translatable. You can use this : `self.env._("Only one offer can be accepted per property.")`

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,54 @@
+                lambda o: o.status == "accepted" and o.id != record.id
+            ):
+                raise UserError("Only one offer can be accepted per property.")
+            record.status = "accepted"
```
Don't forget about using single quotes here as it is a technical string.

lost-odoo commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,15 @@
+
+    _check_unique_name = models.Constraint(
+        "UNIQUE(name)", "The tag name must be unique."
+    )
```
Same here on a single line :smile:

lost-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,14 @@
+    _order = "name"
+    _check_unique_name = models.Constraint(
+        "UNIQUE(name)", "The type name must be unique."
+    )
```
Here too

lost-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,35 @@
+              parent="estate_settings_menu" 
+              action="estate_property_tag_action"/>
+
+      <menuitem id="estate_property_offer_menu_action" 
```
Seems like there is a spacing issue here :smile:

lost-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,46 @@
+<odoo>
+    <!-- Tree View -->
+    <record id="estate_property_offer_view_tree" model="ir.ui.view">
+        <field name="name">estate.property.offer.view.tree</field>
```
Try to use list instead of tree . I know that you can find in examples `tree` but it is because before it was named tree and was changed to list.
Same in other views.

lost-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,110 @@
+        <field name="res_model">estate.property</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
Here too, don't forget about the final new line.

LoganStaelens replied:
Same

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+    )
+    _check_selling_price = models.Constraint(
+        "Check(selling_price >= 0)", "The selling price must be positive."
+    )
```
Here too, it can be on one line each.

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+    )
+
+    name = fields.Char(required=True)
+    description = fields.Text()
```
Try to at least put the string parameter to every fields if possible. Even if the name of the field is implied.

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+                    < 0
+                ):
+                    raise ValidationError(
+                        "The selling price cannot be lower than 90% of the expected price."
```
Here also it should be translated.

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+    def action_sold(self):
+        for record in self:
+            if record.state == "cancelled":
+                raise UserError("Cancelled property cannot be sold.")
```
Same here

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+    def action_cancel(self):
+        for record in self:
+            if record.state == "sold":
+                raise UserError("Sold property cannot be cancelled.")
```
And here

lost-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,106 @@
+                        precision_digits=2,
+                    )
+                    < 0
+                ):
```
Can you put this on one line, it shouldn't be too long.
```suggestion
                if float_compare(record.selling_price, record.expected_price * 0.9, precision_digits=2) < 0:
```

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,30 @@
+        'views/estate_property_offer_views.xml',
+        'views/estate_menus.xml',
+    ],
+    'installable': True,
```
`installable` is not useful as `application` is set to true.

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,30 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': 'Real Estate Advertisement',
+    'version': '1.0',
```
You don't have to provide the version because it is `1.0` by default.

lost-odoo commented on README-raibr-docs.md:
```diff
@@ -0,0 +1 @@
+My first change on odoo
```
Is it still useful to keep that ? :thinking:

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,57 @@
+        string="Deadline",
+    )
+
+    # self.property_id.status = 'offer_received' when offer is created
```
This is useless as we can easily understand via the code right below :smile:

lost-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,57 @@
+    _order = 'price desc'
+
+    price = fields.Float(required=True, string="Price")
+    _check_price = models.Constraint('Check(price > 0)', "The offer price must be strictly positive.")
```
Try to avoid mixing fields with constraints so it is more readable.

lost-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,19 @@
+    property_ids = fields.One2many(comodel_name='estate.property', inverse_name='property_type_id', string="Properties")
+    offer_count = fields.Integer(compute='_compute_offer_count', string="Offer Count")
+
+    def _compute_offer_count(self):
```
Is it normal that you don't have an `api.depends` here ?

ramezlahzy replied:
that function is called when that field calculated 
    offer_count = fields.Integer(compute='_compute_offer_count', string="Offer Count")
    so, there is no need for any dependency

LoganStaelens replied:
Yeah but I think you still need to add a dependency as if property_ids change it is not recomputed.

lost-odoo commented on estate/views/estate_property_type_views.xml:
```diff
@@ -0,0 +1,52 @@
+        <field name="res_model">estate.property.type</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
Please don't forget about the final new line.

LoganStaelens replied:
Same here

lost-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,123 @@
+        <field name="view_mode">list,form</field>
+        <field name="context">{'search_default_filter_available': True}</field>
+    </record>
+</odoo>
```
Hre too

lost-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,53 @@
+        <field name="domain">[('property_id.property_type_id', '=', active_id)]</field>
+        <field name="domain">[('property_id.property_type_id', '=', active_id)]</field>
+    </record>
+</odoo>
```
Here also

lost-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,53 @@
+        <field name="res_model">estate.property.offer</field>
+        <field name="view_mode">list,form</field>
+        <field name="domain">[('property_id.property_type_id', '=', active_id)]</field>
+        <field name="domain">[('property_id.property_type_id', '=', active_id)]</field>
```
Duplicate lines here :smile:

lost-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,53 @@
+                <field name="status" />
+                <button name="action_accept" type="object" icon="fa-check" title="Accept" />
+                <button name="action_refuse" type="object" icon="fa-times" title="Refuse" />
+
```
Blank lines

ramezlahzy committed:
[IMP] estate: Implement salesperson management with properties association

ramezlahzy committed:
[IMP] estate_account (Chapter 13): Add Real Estate Accounting integration with automatic invoice generation

ramezlahzy committed:
[IMP] estate (Chapter 14): Add Kanban view for properties and associated styles

ramezlahzy committed:
[IMP] estate: fix some linting problem

LoganStaelens commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,100 @@
+    @api.constrains('selling_price', 'expected_price')
+    def _check_selling_price(self):
+        for property in self:
+            if not float_is_zero(property.selling_price, precision_digits=2):
+                if float_compare(property.selling_price, property.expected_price * 0.9, precision_digits=2) < 0:
```
Why doing two conditions ?
```suggestion
            if not float_is_zero(property.selling_price, precision_digits=2) and float_compare(property.selling_price, property.expected_price * 0.9, precision_digits=2) < 0:
```

LoganStaelens commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,57 @@
+
+    def action_accept(self):
+        for offer in self:
+            if offer.property_id.offer_ids.filtered(lambda o: o.status == 'accepted' and o.id != offer.id):
+                raise UserError(self.env._("Only one offer can be accepted per property."))
```
Can you not check the state of the property directly, instead of doing a filtering ?

LoganStaelens reviewed (changes requested):
Did another review.

LoganStaelens commented on estate/static/src/css/estate.css:
```diff
@@ -0,0 +1,18 @@
+/* Estate Module Kanban View Styles */
+
+.o_kanban_record_title {
+    font-size: 16px;
+    font-weight: bold;
+    margin-bottom: 8px;
+    color: #212529;
+}
+
+.o_kanban_record_subtitle {
+    font-size: 13px;
+    color: #6c757d;
+    margin-bottom: 4px;
+}
+
+.o_kanban_tags_section {
+    margin-top: 8px;
+}
```
Try to at least put your module name to avoid collision with the rest of the css from odoo

LoganStaelens commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,46 @@
+                    'partner_id': property.buyer_id.id,
+                    'move_type': 'out_invoice',
+                    'invoice_line_ids': [
+                        (
+                            0,
+                            0,
+                            {
+                                'name': 'Property price',
+                                'quantity': 1,
+                                'price_unit': property.selling_price,
+                            },
+                        ),
```
Here you could use Command.create from odoo.

LoganStaelens commented on estate_account/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    "data": [],
+    "application": False,
+    "license": "LGPL-3",
+}
```
Don't forget to add auto_install here so your bridge module is auto installed when all dependencies are installed.
